File: gtrace/draw/viewer/widget.js
/*
 * anywidget front end for the gtrace viewer (Stage 2).
 *
 * This file is not standalone: widget.py concatenates viewer.js in front
 * of it and serves the result as the widget's ESM module. viewer.js is
 * an IIFE that publishes GTraceViewer on globalThis, so the core is
 * already in scope by the time render() runs. Everything specific to
 * the notebook transport lives here, so that the same core can be
 * driven by a static HTML page (Stage 1) or a websocket (Stage 3)
 * without change.
 *
 * Copyright (c) 2011-2026, Yoichi Aso. BSD license.
 */

/*
 * How tall the viewer makes itself when nothing has said.
 *
 * A cell output is a letterbox, and 520 pixels of one is not enough of a
 * bench drawing to work in: the first thing anyone did with it was drag
 * it taller. So the height is taken from the width, which is the cell's
 * own and therefore already the right scale for the window.
 *
 * From the width of the *drawing*, though, not of the widget: the side
 * panel is a fixed 380 pixels of it and is a column of numbers, not part
 * of the picture. Squaring the whole widget makes the drawing itself
 * taller than it is wide, which is the wrong way round for a bench.
 *
 * Capped, because a maximized JupyterLab cell is wider than the window
 * is tall and a viewer whose bottom edge is below the fold is unusable
 * in its own way. The cap never falls below FALLBACK_HEIGHT: an
 * embedder that reports a nonsense viewport - an output frame measuring
 * only what it has so far been given, say - should leave the widget at
 * the size it used to be rather than at nothing.
 */
const SIDE_PANEL_WIDTH = 380;
const NARROW_BREAKPOINT = 700;
const VIEWPORT_FRACTION = 0.70;
const FALLBACK_HEIGHT = 520;

function autoHeight(el) {
    const width = el.getBoundingClientRect().width;
    // Nothing to go on yet. anywidget calls render() before the output
    // area is laid out, so this is the ordinary case on the first pass,
    // not an error: the observer below runs again when a width arrives.
    if (!width) { return 0; }
    // Narrow enough and the side panel stops standing beside the
    // drawing and stacks under it - the same breakpoint the stylesheet
    // uses - so there the drawing has the whole width.
    const drawing = width > NARROW_BREAKPOINT
        ? width - SIDE_PANEL_WIDTH : width;
    const cap = Math.max((globalThis.innerHeight || 0) * VIEWPORT_FRACTION,
                         FALLBACK_HEIGHT);
    const floor = globalThis.GTraceViewer.MIN_HEIGHT;
    return Math.max(floor, Math.round(Math.min(drawing, cap)));
}

function render({ model, el }) {
    const host = document.createElement('div');
    host.className = 'gt-widget';

    // A height of zero means "work it out": see autoHeight. Anything
    // else is a height someone chose, from Python or by dragging the
    // grip, and is used as it stands.
    //
    // Until a width is known there is nothing to work it out from, and
    // the fixed height the widget used to have stands in. That is a
    // provisional answer rather than a fallback to live with: whichever
    // of the three chances below sees a real width first replaces it,
    // and refits, since the drawing was framed against the wrong box.
    let resolved = false;
    const applyHeight = () => {
        const set = model.get('height');
        if (set) {
            host.style.height = set + 'px';
            return true;
        }
        const auto = autoHeight(el);
        host.style.height = (auto || FALLBACK_HEIGHT) + 'px';
        if (!auto || resolved) { return !!auto; }
        // The first time a real width is known. Anything drawn before
        // now was framed against the provisional height.
        resolved = true;
        if (el.gtraceViewer) { el.gtraceViewer.fit(); }
        return true;
    };
    applyHeight();
    el.appendChild(host);

    // Editing is enabled by handing the core somewhere to send edits.
    // Here that is the widget's comm; the live server will hand it a
    // websocket instead, with the same message format.
    const onEdit = model.get('editable')
        ? (msg) => model.send(msg)
        : null;

    // Set from Python, a new height reframes the drawing to suit it.
    // Set by a drag of the bottom edge, it must not: the view is already
    // where the user put it, and refitting would throw away the zoom
    // they were working at.
    let dragged = false;

    // A cell output is a letterbox and a bench drawing is not, so the
    // viewer can be dragged taller by its bottom edge. The height is
    // this element's, so it is written back to the traitlet that set it:
    // that is what makes the new height survive a re-render, and lets
    // Python see what it was dragged to.
    const viewer = globalThis.GTraceViewer.mount(host, model.get('scene'), {
        layoutPath: model.get('layout_path'),
        dxfPath: model.get('dxf_path'),
        onEdit: onEdit,
        resizable: true,
        onResize: (h) => {
            dragged = true;
            model.set('height', h);
            model.save_changes();
        }
    });

    // Python pushes a new scene whenever the layout is re-traced. Keep
    // the current zoom, pan and layer visibility so that a re-trace does
    // not throw away where the user was looking.
    const onScene = () => viewer.setScene(model.get('scene'));
    const onHeight = () => {
        applyHeight();
        if (dragged) { dragged = false; return; }
        viewer.fit();
    };

    // While the height is being worked out, it follows the width: split
    // the notebook pane and the viewer squares itself up again. It stops
    // following the moment anything sets a height - the grip writes one
    // back, so a drag settles it for good.
    //
    // This is also what resolves the first pass. anywidget calls
    // render() before the output area has been laid out, so the width
    // there is zero; the run that matters is the one this fires when a
    // real width arrives. Getting a fit wrong at that moment is what
    // once left the whole drawing scaled down to a point, so the height
    // is only ever applied here - the viewer refits itself off its own
    // observer once it has a size.
    let observer = null;
    if (globalThis.ResizeObserver) {
        observer = new ResizeObserver(() => {
            if (!model.get('height')) { applyHeight(); }
        });
        observer.observe(el);
    }

    // Two more chances, because the observer is not to be relied on for
    // the one run that matters. A resize callback is delivered on a
    // rendering step, and an embedder that has already laid the output
    // area out before calling render() never resizes it again - there
    // is nothing for the observer to report. A headless browser stops
    // delivering them altogether once the page has settled, which is
    // how this was found. Both of these are cheap and idempotent.
    const retry = () => { if (!model.get('height')) { applyHeight(); } };
    if (globalThis.requestAnimationFrame) { requestAnimationFrame(retry); }
    const retryTimer = setTimeout(retry, 0);

    // An edit Python refused: say so rather than leaving the drawing
    // silently disagreeing with what the user just did.
    const banner = document.createElement('div');
    banner.className = 'gt-error';
    banner.style.display = 'none';
    el.appendChild(banner);
    const onError = () => {
        const msg = model.get('error');
        banner.textContent = msg;
        banner.style.display = msg ? '' : 'none';
        // A refused edit sends no new scene, so anything the viewer
        // assumed about it - a rename it followed ahead of Python -
        // has to be put back.
        if (msg) { viewer.revertSelection(); }
    };

    // Confirmation of something that leaves no mark on the drawing.
    const notice = document.createElement('div');
    notice.className = 'gt-notice';
    notice.style.display = 'none';
    el.appendChild(notice);
    const onNotice = () => {
        const msg = model.get('notice');
        notice.textContent = msg;
        notice.style.display = msg ? '' : 'none';
    };

    model.on('change:scene', onScene);
    model.on('change:height', onHeight);
    model.on('change:error', onError);
    model.on('change:notice', onNotice);
    onError();
    onNotice();

    // Expose the viewer for debugging and for the tests. The height
    // resolver goes with it: what it comes to depends on the width of
    // an element and on the size of the window, neither of which a
    // check can arrange and then wait on reliably in a headless run.
    el.gtraceViewer = viewer;
    el.gtraceApplyHeight = applyHeight;

    return () => {
        model.off('change:scene', onScene);
        model.off('change:height', onHeight);
        model.off('change:error', onError);
        model.off('change:notice', onNotice);
        if (observer) { observer.disconnect(); observer = null; }
        clearTimeout(retryTimer);
        viewer.destroy();
    };
}

export default { render };

File: gtrace/draw/viewer/widget.py
'''
Jupyter widget front end for the gtrace viewer (Stage 2).

LayoutViewer wraps the same dependency-free viewer core that the static
HTML output uses, so a notebook gets zoom, pan and click readout without
leaving the cell. The scene travels over the widget's traitlets, which
means a re-trace can be pushed into a live view:

    w = layout.widget()
    w                      # displays the viewer
    M1.HRcenter = [0.6, 0]
    w.update()             # re-traces and redraws in place

anywidget is an optional dependency: importing gtrace does not require
it, and the HTML output keeps working without it.
'''

#{{{ Import modules

from gtrace.draw.viewer import viewer_js, viewer_css, _read_asset

#}}}

#{{{ Author and License Infomation

#Copyright (c) 2011-2026, Yoichi Aso
# All rights reserved.

# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
# * Redistributions of source code must retain the above copyright
#   notice, this list of conditions and the following disclaimer.
#
# * Redistributions in binary form must reproduce the above copyright
#   notice, this list of conditions and the following disclaimer in the
#   documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

__author__ = "Yoichi Aso"
__copyright__ = "Copyright 2011-2026, Yoichi Aso"
__credits__ = ["Yoichi Aso"]
__license__ = "BSD"
__version__ = "0.3.0"
__maintainer__ = "Yoichi Aso"
__email__ = "yoichi.aso@nao.ac.jp"
__status__ = "Beta"

#}}}

#{{{ Availability

class WidgetNotAvailable(ImportError):
    '''
    Raised when the notebook viewer is asked for but anywidget is
    not installed.
    '''
    pass

def widget_available():
    '''
    Whether the notebook viewer can be used, i.e. whether anywidget is
    importable.
    '''
    try:
        import anywidget  # noqa: F401
    except ImportError:
        return False
    return True

def _require_anywidget():
    try:
        import anywidget
        import traitlets
    except ImportError:
        raise WidgetNotAvailable(
            'The notebook viewer needs anywidget. Install it with '
            "'pip install anywidget' (or 'pixi add anywidget'), or use "
            'layout.show(backend="html") to open the viewer in a browser '
            'instead.')
    return anywidget, traitlets

#}}}

#{{{ ESM module

def widget_js():
    '''
    Return the source of the anywidget ESM wrapper as a string.
    '''
    return _read_asset('widget.js')

def widget_esm():
    '''
    Return the ESM module served to the notebook front end.

    It is the viewer core followed by the anywidget wrapper. The core is
    an IIFE that publishes GTraceViewer on globalThis, so concatenating
    the two is enough to bring it into scope; there is no build step and
    no module resolution to go wrong, which was the lesson of the
    dxf-viewer experiment.
    '''
    return viewer_js() + '\n' + widget_js()

#}}}

#{{{ LayoutViewer

_LAYOUT_VIEWER = None

def _build_class():
    '''
    Define the widget class on first use, so that importing this module
    does not require anywidget.
    '''
    anywidget, traitlets = _require_anywidget()

    class LayoutViewer(anywidget.AnyWidget):
        '''
        Notebook viewer for a gtrace scene.

        Parameters
        ----------
        scene : dict
            A scene dict as returned by serialize.scene_to_dict().
        title : str
            Title shown in the side bar.
        height : int
            Height of the viewer in pixels.
        '''

        _esm = widget_esm()
        _css = viewer_css()

        scene = traitlets.Dict().tag(sync=True)
        title = traitlets.Unicode('gtrace').tag(sync=True)
        height = traitlets.Int(520).tag(sync=True)
        #: Whether the front end may send edit messages back.
        editable = traitlets.Bool(True).tag(sync=True)
        #: Last edit failure, shown by the front end. Empty when fine.
        error = traitlets.Unicode('').tag(sync=True)
        #: Confirmation of something that leaves no visible trace, such
        #: as a save. Empty when there is nothing to say.
        notice = traitlets.Unicode('').tag(sync=True)
        #: File the Save and Load buttons start on.
        layout_path = traitlets.Unicode('layout.json').tag(sync=True)

        def __init__(self, scene=None, layout=None, draw_kwargs=None,
                     **kwargs):
            super().__init__(scene=scene if scene is not None else {},
                             **kwargs)
            # Held so that update() can re-trace. Not a traitlet: the
            # layout is a Python object, not something to synchronize.
            self._layout = layout
            self._draw_kwargs = dict(draw_kwargs or {})
            self._edit_log = []
            if layout is None:
                self.editable = False
            # Note the name: Widget._handle_custom_msg(content, buffers)
            # is ipywidgets' own dispatcher, and overriding it with a
            # callback signature breaks every custom message.
            self.on_msg(self._on_edit_msg)

        def update(self, **kwargs):
            '''
            Re-trace the layout this viewer came from and redraw, keeping
            the current zoom, pan and layer visibility.

            Parameters
            ----------
            **kwargs
                Passed to OpticalLayout.draw(). Remembered, so that
                edits coming from the front end draw the same way.
            '''
            if self._layout is None:
                raise ValueError('This viewer is not attached to a layout; '
                                 'assign to .scene directly instead.')
            if kwargs:
                self._draw_kwargs = dict(kwargs)
            self._layout.beams = None
            self.scene = self._layout.scene_dict(**self._draw_kwargs)
            return self

#{{{ Edits from the front end

        @property
        def edits(self):
            '''
            The edit messages received so far, oldest first.
            '''
            return list(self._edit_log)

        def apply_edit(self, msg):
            '''
            Apply one edit message to the layout and push the new scene.

            This is what the front end reaches through the widget's
            comm, and is also callable directly, which is how the tests
            drive it. A rejected edit leaves the layout untouched and
            reports itself through the 'error' traitlet rather than
            raising into the comm machinery, where nothing would see it.

            Parameters
            ----------
            msg : dict
                An edit message, as described by
                OpticalLayout.apply_edit.

            Returns
            -------
            ok : bool
                Whether the edit was applied.
            '''
            if self._layout is None or not self.editable:
                self.error = 'This viewer is read-only.'
                return False
            try:
                self._layout.apply_edit(msg)
            except Exception as e:
                self.error = '%s: %s' % (type(e).__name__, e)
                return False
            self._edit_log.append(msg)
            self.error = ''

            # A save leaves the drawing exactly as it was, so without a
            # word from us the user cannot tell it happened.
            if msg.get('op') == 'save':
                self.notice = 'Saved to %s' % (msg.get('path'),)
                return True
            self.notice = ('Loaded %s' % (msg.get('path'),)
                           if msg.get('op') == 'load' else '')
            self.scene = self._layout.scene_dict(**self._draw_kwargs)
            return True

        def _on_edit_msg(self, widget, content, buffers):
            '''
            Callback registered with on_msg. ipywidgets calls it as
            (widget, content, buffers) for every custom message the
            front end sends, so ignore anything that is not an edit.
            '''
            if isinstance(content, dict) and 'op' in content:
                self.apply_edit(content)

#}}}

    return LayoutViewer

def LayoutViewer(*args, **kwargs):
    '''
    Construct the notebook viewer widget.

    This is a function rather than the class itself so that anywidget is
    imported only when a widget is actually created.
    '''
    global _LAYOUT_VIEWER
    if _LAYOUT_VIEWER is None:
        _LAYOUT_VIEWER = _build_class()
    return _LAYOUT_VIEWER(*args, **kwargs)

#}}}
